File: 总题库/1382.BalanceABinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    vector<int> inorderList;
    void getInorder(TreeNode* root) {
        if (root == NULL) return;
        getInorder(root->left);
        inorderList.push_back(root->val);
        getInorder(root->right);
    }

    TreeNode* build(int left, int right) {
        if (right - left < 0) return NULL;
        int middle = (left + right) >> 1;
        TreeNode *root = new TreeNode(inorderList[middle]);
        root->left = build(left, middle - 1);
        root->right = build(middle + 1, right);
        return root;
    }

public:
    TreeNode* balanceBST(TreeNode* root) {
        getInorder(root);
        root = build(0, inorderList.size() - 1);
        return root;
    }
};

File: 总题库/1382.BalanceABinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        def getInorder(root: TreeNode) -> None:
            if root == None:
                return None
            
            getInorder(root.left)
            inorderList.append(root.val)
            getInorder(root.right)

        def build(left: int, right: int) -> TreeNode:
            if right - left < 0:
                return None

            middle = (left + right) // 2
            root = TreeNode(inorderList[middle])
            root.left = build(left, middle - 1)
            root.right = build(middle + 1, right)
            return root

        inorderList = list()
        getInorder(root)
        root = build(0, len(inorderList) - 1)
        return root